text/0000-deprecate-array-prototype-extensions.md

```diff
@@ -0,0 +1,178 @@
+    this.abc.pushObject(newItem);
+  }
+};
+```
+
+Examples of current code. 
+#### Option 1: use `tracked` property
+```js
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+
+export default class SampleComponent extends Component{
+  @tracked abc = [];
+
+  get lastObj() {
+    return this.abc.at(-1);
+  }
+
+  @action
+  someAction(newItem) {
+    this.abc = [...abc, newItem];
+  }
+};
+```
+
+#### Option 2: use `TrackedArray`
+```js
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { TrackedArray } from 'tracked-built-ins';
+
+export default class SampleComponent extends Component{
+  abc = new TrackedArray();
+
+  get lastObj() {
+    return this.abc.at(-1);
+  }
+
+  @action
+  someAction(newItem) {
+    abc.push(newItem);
+  }
+};
+```
+
+### Observable properties in templates
+Examples of deprecated code:
+
+```hbs
+<Foo @bar={{@list.firstObject.name}} />
+```
+
+Examples of current code:
+```hbs
+<Foo @bar={{get @list '0.name'}} />
+```
+
+After the deprecated code is removed from Ember (at 5.0), we need to remove the [options to disable the array prototype extension](https://guides.emberjs.com/v4.2.0/configuring-ember/disabling-prototype-extensions/) from Official Guides and we also need to update the [Tracked Properties](https://guides.emberjs.com/v4.2.0/upgrading/current-edition/tracked-properties/#toc_arrays) `Arrays` section with updated suggestions.
+
+## Drawbacks
+- For users relying on Ember array prototype extensions, they will have to refactor their code and use equivalences appropriately.
+- A lot of existing Ember forums/blogs had been assuming the enabling of array prototype extensions which could cause confusions for users referencing them.
+- Increase package sizes, for example, before `this.abc.filterBy('x');`, now `this.abc.filter(el => el !== 'x');`.
+- Although `tracked-built-ins` is on the path to stabilization as an official API via [RFC #812](https://github.com/emberjs/rfcs/pull/812), it is not yet officially recommended and its API may change somewhat between now and stabilization.
+
+## Alternatives
+- Do the deprecation as suggested here for use within Ember itself, but extract it as a standalone library for users who want to use it. This will only work as long as the underlying Ember Classic reactivity system is supported.
+
+    As a variation on this, we could do this but explicitly only support it up through the first LTS release in the 5.x series.
+
+- Continuing allowing array prototype extensions but turning the `EXTEND_PROTOTYPES` off by default.
+
+- Do one of these, but target Ember v6 instead.
+
+- Do nothing.
+
+## Unresolved questions
+- Current lint rule [ember/no-array-prototype-extensions](https://github.com/ember-cli/eslint-plugin-ember/blob/master/docs/rules/no-array-prototype-extensions.md) gives a lot of false positives giving the limitation of static analysis.
```

Comment:
these false positives will be greatly reduced soon because in ember-data 4.8 usages of these has already been deprecated. This means remaining usage is from `A()` which I would hope this RFC would consider simultaneously deprecating or various other `ArrayProxy` implementations which are far less common.

Reply:
Agreed! One of the things I plan to discuss at Friday's Framework Core meeting is exactly whether we should deprecate `@ember/array` entirely and/or *at least* extracting it and move it out to a separate package which is not installed by default.

Reply:
Sure, but the reason I mention `A` is that `A` mutates array instances when prototype extensions are turned off. If we're going to rip off the bandaide it seems better to rip it fully off vs allowing people to just call `A` on everything to get back to the same state.

Reply:
Another sets of false positives are self-defined class functions, for example`invoke`, `clear` are pretty common method names. ☹️

Reply:
@runspired yeah, I spent literally my whole day today dealing specifically with the horror caused by the `NativeArray` shenanigans as relates to types. If it were my preference alone, we'd deprecate it and cut a special v5.0 *tomorrow* just to get rid of `A()`. :joy:

@smilland yeah, unfortunately that kind of thing is basically impossible to avoid 100% without full TS coverage (which we obviously can't require people to have!).